.github/workflows/permit-generator.yml

```diff
@@ -0,0 +1,49 @@
+name: Permit Generator
+
+on:
+  workflow_dispatch:
+    inputs:
+      users_amounts:
+        description: "A JSON array containing usernames and associated amounts"
+        required: true
+        # example: '[{"user1": 100}, {"user2": 150}]'
+
+jobs:
+  run:
+    runs-on: ubuntu-latest
+    permissions: write-all
+
+    steps:
+      - uses: actions/checkout@v4
+
+      - name: Set up Node.js
+        uses: actions/setup-node@v4
+        with:
+          node-version: "20.10.0"
+
+      - name: Install dependencies
+        run: yarn install --immutable --immutable-cache --check-cache
+
+      - name: Process permit requests
+        id: parse
+        run: |
+          echo "Received input: ${{ github.event.inputs.users_amounts }}"
+          USERS_AMOUNTS=${{ github.event.inputs.users_amounts }}
+          echo "$USERS_AMOUNTS" > users_amounts.json
+
+          cat users_amounts.json | jq -c '.[]' | while read user_amount; do
+            USERNAME=$(echo $user_amount | jq -r 'keys[0]')
+            AMOUNT=$(echo $user_amount | jq -r '.[]')
+
+            # TODO: This will convert the username to it's unique user ID.
+            USER_ID=$(npx tsx scripts/convertUsernameToID.ts $USERNAME)
+
+            #TODO: Generate a permit for the user using the user ID and the amount provided
+            PERMIT=$(npx tsx scripts/generate-permit.ts $USER_ID $AMOUNT)
+
+            # TODO: Implement logic for how to output the permit to the user.
```

Comment:
Please give me more info about this one.

Reply:
We just need to reply so that we can receive it in the kernel @gentlementlegen @whilefoo how do you think is best?

Reply:
I think that this plugin should not be able to be triggered by anyone or anyone can generate permits of any amount to anywho. Also we can use our permit package that is already covering all of this and have it running inside this plugin. Not sure what benefit we get over just importing and using the package where needed.

Reply:
The original planned architecture is centralized permit generation. 

We need to take a fee. It's part of our business model. 

Permit generation locally was a shortcut done to get plugins working faster.

Reply:
Sure, just we should find of a proper system to specify which wallet is used and how to convey the data safely.

Reply:
> specify which wallet is used

I suppose our centralized database for now.

Reply:
Well this kinda goes against the rollup feature, because we will be storing rewards/penalties in the database and generating a permit on-demand when the user wants to claim their rewards.

If the vision is that other third-party plugins also generate permits, then are they going to write to our DB with rewards and then fee is taken when the user claims on `pay.ubq.fi` or will they generate permits with `permit-generation` and handle them in their own way?

In the second scenario, this repo can just be an Action plugin that generates a permit and takes a fee like we originally wanted, so you can have a plugin chain with first plugin being the third party plugin that uses their own algorithm to calculate rewards and output them in a format that `permit-generation` plugin wants.

Reply:
Hmm I'm not 100% on implementation details right now but understand that permit generation must be centralized or we can't enforce taking a fee. Perhaps we should discuss specific proposals on how to implement.

Reply:
> or we can't enforce taking a fee

am I correctly understanding the fee concept as if the reward is 50$, we take for example 1$ and the user gets 49$?

> I think that the kernel should compile all rewards outputs from all plugins per webhook event and then request permits at the end.
> 
> If they are 0 then don't request a permit automatically. Not every webhook event is to generate a reward (like label changes.) handling this automatically will make the developer experience simpler. We just need an SDK method to be able to return requested rewards to the kernel.

The problem with this approach is that the plugin that generated the rewards can't get the permits back from the kernel because it happens after all plugins are already finished executing, so they can't post a comment with permit or any other way they want.

We could combine this approach with the rollup feature so that all plugins tell the kernel the rewards and at the end the kernel stores rewards in the DB and user can claim on `pay.ubq.fi`.

Reply:
> > or we can't enforce taking a fee
> 
> 
> 
> am I correctly understanding the fee concept as if the reward is 50$, we take for example 1$ and the user gets 49$?
> 

Yes

> 
> 
> > I think that the kernel should compile all rewards outputs from all plugins per webhook event and then request permits at the end.
> 
> > 
> 
> > If they are 0 then don't request a permit automatically. Not every webhook event is to generate a reward (like label changes.) handling this automatically will make the developer experience simpler. We just need an SDK method to be able to return requested rewards to the kernel.
> 
> 
> 
> The problem with this approach is that the plugin that generated the rewards can't get the permits back from the kernel because it happens after all plugins are already finished executing, so they can't post a comment with permit or any other way they want.
> 

Situations like this I am proposing that we accommodate by hard coding stuff in the kernel to facilitate. It's not the most elegant approach but it's a viable one. 

> 
> 
> We could combine this approach with the rollup feature so that all plugins tell the kernel the rewards and at the end the kernel stores rewards in the DB and user can claim on `pay.ubq.fi`.
> 
> 
> 
> 

That sounds good to me. I think that roll ups require a bit more trust of the system compared to regular permits. I can see how permits might be appealing to some organizations.

Reply:
As far as I understand we must "centralize" permit generation (i.e. partners shouldn't be able to fork https://github.com/ubiquibot/conversation-rewards plugin and generate permits themselves).

It makes sense to convert the https://github.com/ubiquibot/permit-generation npm package to a cloudflare worker instead of a github action "plugin" because:
1. It fits with the permit rollups (i.e. generate permits on demand). This way `pay.ubq.fi` backend will have to call the `permit-generation` URL to get a signed permit.
2. It fits with the `conversation-rewards` plugin because all we have to do is to refactor it to get a signed permit from a URL instead of an npm package.

I suppose both `conversation-rewards` plugin and `pay.ubq.fi` backend will have to provide some API key in order to prohibit anybody from signing permits.

Regarding plugin developer rewards I always anticipated it working this way:
1. Plugin dev sets a fee in a plugin's `manifest.json`
2. `conversation-rewards` plugin generates an additional fee permit (or saves an additional "reward" in terms of a rolled permites up feature) besides the  original github issue reward permit

@gentlementlegen @whilefoo If it sounds good then I'll updated description for https://github.com/ubiquibot/permit-generation/issues/68.

Reply:
Fine by me, let's just make sure that Cloudflare implementation of `npm:crypto` works properly because a lot of `node` native functions are mocked within Workers, or are experimental. Here is the compatibility list:
https://developers.cloudflare.com/workers/runtime-apis/nodejs/crypto/

Reply:
> I suppose both `conversation-rewards` plugin and `pay.ubq.fi` backend will have to provide some API key in order to prohibit anybody from signing permits.

That would mean the any third party plugins that want to generate permits will also need the API key and could generate permits whenever they wanted. I think it would be better to restrict plugins so they can only generate permits when the plugin is triggered by an event

Reply:
> That would mean the any third party plugins that want to generate permits

Why would 3rd party plugins want to generate payout permits? Are there any real use cases? It's kind of strange to allow access to partners' wallet private key (by allowing to generate permits) for 3rd parties.

Reply:
I am hoping we can get by without API keys and instead inherit authentication from GitHub or whatever service the request originates from. 

I imagine there must be some way we can verify authentication based on the bot installation and the GitHub repo. 

We could do some research and iterate our authentication as we get further along.

Reply:
> 4\. The kernel should request these permits to be generated from this api. Because it's our kernel I think that we should be able to verify that it's the kernel somehow

We can basically run it as a Worker plugin, because we can verify it's our kernel because the kernel signs the payload and the plugin verifies it with kernel's public key

Reply:
Ok, as far as I understand we keep the `https://github.com/ubiquibot/permit-generation` plugin as a github action.

@whilefoo 

> we can verify it's our kernel because the kernel signs the payload and the plugin verifies it with kernel's public key

A few questions:
1. "kernel signs the payload" is already implemented, correct?
2. `ubiquibot/permit-generation` plugin doesn't verify the caller, correct? I mean this functionality is not implemented right now.
3. Are there any plugins which verify a caller? If so could you share a related code?

Reply:
> 1. "kernel signs the payload" is already implemented, correct?

Yes



> 2\. `ubiquibot/permit-generation` plugin doesn't verify the caller, correct? I mean this functionality is not implemented right now.

Well technically anybody with workflow permission on the repo can call it but it doesn't verify it's necessarily the kernel

> 3\. Are there any plugins which verify a caller? If so could you share a related code?

Yes, all Worker plugins [verify the caller](https://github.com/ubiquity-os-marketplace/assistive-pricing/blob/9984131a66eb9167dd0f60147a99e41ae382c43b/src/worker.ts#L45)